### PR TITLE
Add LEP for disk anti-affinity

### PR DESCRIPTION
**Pull request by ejweber**
#3823

### CODE REVIEW

File: enhancements/20230718-disk-anti-affinity.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  - Increased data durability. If a single disk fails, a healthy replica will still exist on an disk that
+    has not failed.
+  - Increased data availability. If a single disk on a node becomes unavailable, but the node itself remains
+    healthy, at least one replica remains healthy. On a single-node cluster, this can directly prevent a volume crash.
+    On a small, multi-node cluster, this can prevent a future volume crash due to the loss of a different node.
+
+### Goals
+
+- In all situations, the Longhorn replica scheduler will make a best effort to ensure two replicas for the same volume
+  do not schedule to the same disk.
```
Comment (james-munson) [inline enhancements/20230718-disk-anti-affinity.md:53]: Kudos to you for this.

File: enhancements/20230718-disk-anti-affinity.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+available disks as desired replicas, I want scheduling to fail obviously. It is important that I know high-priority
+volume isn't being protected so I can take action.
+
+### User Experience In Detail
+
+### API changes
+
+Introduce a new Replica Disk Level Soft Anti-Affinity setting with the following definition. By default, set it to
+`true`. While it is generally desirable to schedule replicas to different disks, it would break with existing behavior
+to refuse to schedule replicas when different disks are not available.
```
Comment (james-munson) [inline enhancements/20230718-disk-anti-affinity.md:86]: I personally think this should default to "false", but I understand the arguments for leaving it as is.  At least the chances are reduced of falling into it accidentally.

**Review by james-munson — APPROVED** [review]
A good step forward.

File: enhancements/20230718-disk-anti-affinity.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+1. In a cluster that includes nodes with multiple available disks, create a volume with
+   `spec.replicaSoftAntiAffinity = true`, `spec.replicaDiskSoftAntiAffinity = true`, and `numberOfReplicas` equal to the
+   total number of disks in the cluster. Confirm that each replica schedules to a different disk. It may be necessary
+   to tweak additional factors. For example, ensure that one disk has enough free space that the old scheduling
+   behavior would assign two replicas to it instead of distributing the replicas evenly among the disks.
+1. In a cluster that includes nodes with multiple available disks, create a volume with
+   `spec.replicaSoftAntiAffinity = false`, `spec.replicaDiskSoftAntiAffinity = true`, and `numberOfReplicas` equal to one more than the total number of disks in the cluster. Confirm that a replica fails to schedule. Previously,
+   multiple replicas would have scheduled to the same disk and no error would have occurred.
+
```
Comment (innobead) [inline enhancements/20230718-disk-anti-affinity.md:147]: `spec.replicaSoftAntiAffinity = false` means each replica should be scheduled to a different node, so `multiple replicas would have scheduled to the same disk and no error would have occurred` is not right, correct?

What you want to test should be `spec.replicaSoftAntiAffinity = true, spec.replicaDiskSoftAntiAffinity = false`?

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just one comment needs to clarify.

File: enhancements/20230718-disk-anti-affinity.md (hunk 3 above)
Comment (ejweber) [reply enhancements/20230718-disk-anti-affinity.md:148]: You are correct. Good catch. Fixed in an amend.

**Review by innobead — APPROVED** [review]

**Pull request merged**